code/datums/components/jam_receiver.dm

```diff
@@ -35,11 +35,37 @@
 	if (move_relay.depth == 0)
 		move_relay.RemoveComponent()
 		associated_relay = null
+	remove_self_from_list(associated_z)
 
 /datum/component/jam_receiver/proc/on_move(...)
 	SIGNAL_HANDLER
+	check_z_list_sanity()
 	check_jammed()
 
+/// gets z of the parent, but possibly not as 0
+/datum/component/jam_receiver/proc/get_sanitized_z()
+	var/atom/parent_atom = parent
+	while(parent_atom.z == 0 && !isarea(parent_atom.loc)) // we try to find a true Z value
+		parent_atom = parent_atom.loc
+	return parent_atom.z
+
+/datum/component/jam_receiver/proc/check_z_list_sanity()
+	var/current_z = get_sanitized_z()
+	if(current_z != associated_z) // z is changed. we change this.
+		remove_self_from_list(associated_z)
+		add_self_to_list(current_z)
+		associated_z = current_z
```

Comment:
I would also just inline this since the name check_z_list_sanity doesn't make sense, since its not checking the sanity of the z-list, its checking if the z value has changed